lms/static/scripts/frontend_apps/errors.js

```diff
@@ -52,3 +65,44 @@ export class APIError extends Error {
     this.details = data.details;
   }
 }
+
+/**
+ * Does the current error object represent an authorization error or need
+ * to re-authorize from the back end?
+ *
+ * This is a special case. We're handling an APIError resulting from an API
+ * request, but there are no further details in the response body to guide us.
+ * This implicitly means that we're facing an authorization-related state.
+ *
+ * Put another way, if an APIError has neither an errorCode nor a serverMessage,
+ * it is considered an "authorization error".
+ *
+ * @param {Error} error
+ * @returns {boolean}
+ */
+export function isAuthorizationError(error) {
+  return error instanceof APIError && !error.serverMessage && !error.errorCode;
+}
+
+/**
+ * Does the current error object represent an API Error with a recognized
+ * backend-provided `errorCode` related to a failed attempt to launch an
+ * assignment?
+ *
+ * @param {Error} error
+ * @returns {error is APIError}
+ */
+export function isLTILaunchServerError(error) {
+  return (
+    error instanceof APIError &&
+    !!error.errorCode &&
+    [
+      'blackboard_file_not_found_in_course',
+      'canvas_api_permission_error',
+      'canvas_file_not_found_in_course',
+      'canvas_group_set_not_found',
+      'canvas_group_set_empty',
+      'canvas_student_not_in_group',
+    ].includes(error.errorCode)
```

Comment:
Wish we could lean on the type for this.

Reply:
Going from the type to a runtime array isn't possible, but in TS 4.5 it will be possible to go the other way: create a list of error codes and then produce a type from that.

```js
const errorCodes = /** @type {const} */ (['code_a', 'code_b', 'code_c']);

// ErrorCode is 'code_a'|'code_b'|'code_c'
/** @typedef {typeof errorCodes[number]} ErrorCode */ 

```

Adapted from the top part of https://stackoverflow.com/a/45257357.